### PR TITLE
exercise threads with log agm calculations

### PR DESCRIPTION
**Pull request by ckormanyos**
The example uses an AGM method to do a "from the ground up" calculation of logarithms. These are then compared with the value from multiprecision's log function.

This example has been motivated in part by discussions in #211

### CODE REVIEW

File: example/exercise_threading_log_agm.cpp [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+    6311, 6317, 6323, 6329, 6337, 6343, 6353, 6359, 6361, 6367, 6373, 6379, 6389, 6397, 6421, 6427, 6449, 6451, 6469, 6473,
+    6481, 6491, 6521, 6529, 6547, 6551, 6553, 6563, 6569, 6571, 6577, 6581, 6599, 6607, 6619, 6637, 6653, 6659, 6661, 6673,
+    6679, 6689, 6691, 6701, 6703, 6709, 6719, 6733, 6737, 6761, 6763, 6779, 6781, 6791, 6793, 6803, 6823, 6827, 6829, 6833,
+    6841, 6857, 6863, 6869, 6871, 6883, 6899, 6907, 6911, 6917, 6947, 6949, 6959, 6961, 6967, 6971, 6977, 6983, 6991, 6997,
+    7001, 7013, 7019, 7027, 7039, 7043, 7057, 7069, 7079, 7103, 7109, 7121, 7127, 7129, 7151, 7159, 7177, 7187, 7193, 7207,
+    7211, 7213, 7219, 7229, 7237, 7243, 7247, 7253, 7283, 7297, 7307, 7309, 7321, 7331, 7333, 7349, 7351, 7369, 7393, 7411,
+    7417, 7433, 7451, 7457, 7459, 7477, 7481, 7487, 7489, 7499, 7507, 7517, 7523, 7529, 7537, 7541, 7547, 7549, 7559, 7561,
+    7573, 7577, 7583, 7589, 7591, 7603, 7607, 7621, 7639, 7643, 7649, 7669, 7673, 7681, 7687, 7691, 7699, 7703, 7717, 7723,
+    7727, 7741, 7753, 7757, 7759, 7789, 7793, 7817, 7823, 7829, 7841, 7853, 7867, 7873, 7877, 7879, 7883, 7901, 7907, 7919
+  }};
```
Comment (NAThompson) [inline example/exercise_threading_log_agm.cpp:212]: This reminds me how badly we need a prime sieve!

**Comment by ckormanyos** [comment]
> This reminds me how badly we need a prime sieve!

Just for exemplary purposes, I will put in an elementary sieve.

This example needs a lot more work such as separatnig the AGM work and reducing or eliminating the time of writing to `cout`. It's a starting point for discussions.

**Comment by jzmaddock** [comment]
>> This reminds me how badly we need a prime sieve!
> Yes. Just for exemplary purposes, I will put in an elementary sieve.

Don't forget we have a big table of primes in Boost.Math already: https://www.boost.org/doc/libs/1_72_0/libs/math/doc/html/math_toolkit/number_series/primes.html

**Comment by ckormanyos** [comment]
>> Yes. Just for exemplary purposes, I will put in an elementary sieve.

> Don't forget we have a big table of primes in Boost.Math already

Ah, yes. I had forgotten. Thank you.
Now using Boost.Math's table of 10,000 `constexpr` primes.
The sieve is in there in a previous Git commit if we ever need it.

Kind regards, Chris

**Comment by ckormanyos** [comment]
So I reduced the impact of writing to `cout` and also changed the digits from 1001 to 501.

I get the following timing (at the moment, I do not have an MPFR around).

```
// 501 digits, 10,000 values, 7 threads
// backend              calculation_time_concurrent      calculation_time_sequential
// cpp_dec_float                   12.3s                            47.8s
// cpp_bin_float                   14.8s                            58.5s
// gmp_float                        1.1s                             4.4s
```

On my benchmarks, there are no spurious preformances, as was mentioned in an off-site preliminary email communication.

I will keep refining and also take a look at MPFR.

A lot of the run-time might come from the log functions and MPFR has, if memory serves, super fast log. So that might be illuminating. I will also look at that.

All in all, however, the thread exercise seems to be moving in the right direction.

**Comment by jzmaddock** [comment]
Great!  Here's what I see:

```
cpp_dec_float:
result_is_ok_concurrent: true, calculation_time_concurrent: 18.1s
result_is_ok_sequential: true, calculation_time_sequential: 48.5s

cpp_bin_float:
result_is_ok_concurrent: true, calculation_time_concurrent: 18.7s
result_is_ok_sequential: true, calculation_time_sequential: 50.4s

gmp:
result_is_ok_concurrent: true, calculation_time_concurrent: 3.3s
result_is_ok_sequential: true, calculation_time_sequential: 12.4s

mpfr:
result_is_ok_concurrent: true, calculation_time_concurrent: 0.6s
result_is_ok_sequential: true, calculation_time_sequential: 1.9s
```

**Comment by jzmaddock** [comment]
@ckormanyos is this good to go?

**Comment by ckormanyos** [comment]
< @ckormanyos is this good to go?

Thank you for asking.
This example has been sitting idle for a few weeks.
Can I get a few more days?

- I would like to ensure that a merge from develop to this branch works.
- Also, is there a place where i could write a (very few) couple of sentences about what this example is doing, its motivation and how it works?

**Comment by ckormanyos** [comment]
> merge from develop to this branch works...

The merge from develop to this branch worked. I forgot unfortunately to tag with CI SKIP. Sorry.

**Comment by ckormanyos** [comment]
The code is ready.
CI passed.
And the comments are now sufficient.

> @ckormanyos is this good to go?

Yes.

**Pull request merged**